Feat/2646 total errors aggregates

jtimpe opened the pull request:
## Summary of Changes
Pull request closes #2646 

* Implements a new `total_errors_by_month` aggregation function, implemented for section 3+4 submissions
* Updates the frontend `SubmissionHistory` component structure to be more readable and limit conditional rendering. some small code duplications and inefficiencies can be ironed out if necessary

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-backend && docker-compose up
cd tdrs-frontend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
1. Submit data files for all sections to test the frontend restructuring
1. Submit data files for sections 3+4 and validate the "Total Errors" column in the submission history

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] A DataFileSummary created for a submission has `case_aggregates` populated with the correct data for the section
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

jtimpe committed:
total_errors_by_month aggregator

jtimpe committed:
impl total_errors aggregator

jtimpe committed:
submission history component structure

jtimpe committed:
total errors frontend impl

jtimpe committed:
Merge branch 'develop' into feat/2646-total-errors-aggregates

jtimpe committed:
add section 3/4 aggregates to tests

jtimpe committed:
Merge branch 'develop' into feat/2646-total-errors-aggregates

jtimpe committed:
lint

jtimpe committed:
fix null coalescing

jtimpe committed:
fix+add aggregates tests

elipe17 commented on tdrs-backend/tdpservice/parsers/test/test_parse.py:
```diff
@@ -887,10 +887,24 @@ def tanf_section3_file(stt_user, stt):
+    dfs.status = dfs.get_status()
+    dfs.case_aggregates = util.total_errors_by_month(
+        dfs.datafile, dfs.status)
+    assert dfs.case_aggregates == {"months": [
```
Is there a way we can fudge one of these files, or make a new one, to capture the case where we have errors across our split out records?

jtimpe replied:
yep! good callout

raftmsohani commented on tdrs-backend/tdpservice/parsers/test/test_parse.py:
```diff
@@ -887,10 +887,24 @@ def tanf_section3_file(stt_user, stt):
+def test_parse_tanf_section3_file(tanf_section3_file, dfs):
     """Test parsing TANF Section 3 submission."""
+    dfs.datafile = tanf_section3_file
+    dfs.save()
```
nit: do you need to save the dfs? Normally the transaction is off so .save() should not matter

jtimpe replied:
they were just holdovers from the old tests. i removed them and the tests still pass, thanks!

raftmsohani commented on tdrs-backend/tdpservice/scheduling/parser_task.py:
```diff
@@ -27,6 +27,8 @@ def parse(data_file_id):
 
     if "Case Data" in data_file.section:
```
not related to this PR but we should have had "Case Data" as a CONSTANT, and I would put it in .DataFile.py

jtimpe commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -358,6 +373,45 @@ def case_aggregates_by_month(df, dfs_status):
     return aggregate_data
+
+
+def total_errors_by_month(df, dfs_status):
```
move to `aggregators.py` once 2711 merges

andrew-jameson replied:
2711 merged earlier this morning, you can now pull in latest from develop

andrew-jameson commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -313,7 +327,8 @@ def month_to_int(month):
 def case_aggregates_by_month(df, dfs_status):
     """Return case aggregates by month."""
     section = str(df.section)  # section -> text
-    program_type = get_prog_from_section(section)  # section -> program_type -> text
+    program_type = get_prog_from_section(
+        section)  # section -> program_type -> text
```
Is this your auto-formatter? Looks odd

andrew-jameson commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -358,6 +373,45 @@ def case_aggregates_by_month(df, dfs_status):
+        rpt_month_year = int(f"{calendar_year}{month_int}")
+
+        for schema_model in schema_models:
+            if isinstance(schema_model, SchemaManager):
+                schema_model = schema_model.schemas[0]
```
What edge case is this catching?

jtimpe committed:
add tribal tests, positive case test

jtimpe committed:
remove unused code from total_errors aggregate func

jtimpe committed:
remove unnecessary saves

jtimpe committed:
Merge branch 'develop' into feat/2646-total-errors-aggregates

jtimpe committed:
move total_errors to aggregates.py

jtimpe committed:
extra blank line

jtimpe committed:
undo formatter changes

jtimpe committed:
extra space

elipe17 reviewed (approved):
Looks good!

raftmsohani commented on tdrs-backend/tdpservice/parsers/aggregates.py:
```diff
@@ -55,3 +55,27 @@ def case_aggregates_by_month(df, dfs_status):
+        month_int = month_to_int(month)
+        rpt_month_year = int(f"{calendar_year}{month_int}")
+
+        error_count = ParserError.objects.filter(
```
we should be able to query all ParserErrors before running the for loop, and have the counts in a table.
 
Another note: as we are parsing more and more files, this query would take forever to run, the problem is df is indexed easily but the rpt_month_year will be slow.

jtimpe replied:
i pulled the main `ParserError.objects.all()` out of the loop - are you suggesting grouping by rpt_month_year or something to that effect as well?

raftmsohani reviewed (changes requested):
more efficient query!

ADPennington commented:
@jtimpe did this fall off my radar?

jtimpe commented:
> @jtimpe did this fall off my radar?

nope! i am still addressing PR feedback and getting approvals - once i've done so i'll follow back up with you

jtimpe committed:
Merge branch 'develop' into feat/2646-total-errors-aggregates

jtimpe committed:
make the query more efficient

raftmsohani reviewed (approved):
LGTM!

andrew-jameson reviewed (approved):
Good to go!

ADPennington committed:
Merge branch 'develop' into feat/2646-total-errors-aggregates

ADPennington committed:
Merge branch 'develop' into feat/2646-total-errors-aggregates

ADPennington commented:
@jtimpe there is a merge conflict here. can you ping me when its ready again?

jtimpe committed:
Merge branch 'develop' into feat/2646-total-errors-aggregates

jtimpe committed:
rm unused code

jtimpe committed:
fix error report url

ADPennington reviewed (approved):
lgtm @jtimpe 🚀 

- Section 3 and 4 total errors for tanf, tribal, and ssp get populated with 0 or non-zero values when status != `Rejected`
- there are a couple of edge cases that I presume will be addressed in later tickets (let me know if this is not the case):
  - `Accepted with errors` status and only error detected is in trailer record: Total Errors == 0 :warning: (maybe [Cat1 Cleanup](https://github.com/raft-tech/TANF-app/pull/2838) will resolve this )
  - any `Partially Accepted with Errors` status: Total Errors == 0 – this status only comes up when at least one data record (not header/trailer) is not correct length to parse. stratum and aggregate data files only have one data record. so this status does not make sense. :warning: will be addressed in [#2754](https://github.com/raft-tech/TANF-app/issues/2754)

test notes [here](https://hackmd.io/P4sQkTRORLGjiodfgRwMtA?view#Feat2646-total-errors-aggregates)
